src/streaming/downsampler.cpp

```diff
@@ -0,0 +1,295 @@
+#include "downsampler.hh"
+#include "macros.hh"
+#include "zarrv2.array.writer.hh"
+#include "zarrv3.array.writer.hh"
+
+namespace {
+template<typename T>
+[[nodiscard]] ByteVector
+scale_image(ConstByteSpan src, size_t& width, size_t& height)
+{
+    const auto bytes_of_src = src.size();
+    const auto bytes_of_frame = width * height * sizeof(T);
+
+    EXPECT(bytes_of_src >= bytes_of_frame,
+           "Expecting at least ",
+           bytes_of_frame,
+           " bytes, got ",
+           bytes_of_src);
+
+    const int downscale = 2;
+    constexpr auto bytes_of_type = static_cast<double>(sizeof(T));
```

Comment:
use of floating point throughout this function is wrong. The calculations here around dimensions should be fixed-point to be correct. Also, prefer fixed width types (int32_t vs int, (u)int64_t vs size_t) where possible. This will make it clearer where you need a certain precision.